### PR TITLE
fix: iv recovery fallback and more

### PR DESCRIPTION
**Pull request by tyrypyrking**
Not so short, but needed. I'm still battling some lingering issues that arose from that data corruption issue. My next PR will likely be targeted at eliminating any possibility of said data corruption during the process of de/encrypting media.

Summary:
A stored 16-byte file iv was trusted unconditionally, so a valid-length-but-wrong iv (decrypting to garbage) left the full image permanently undecodable while its thumbnail (separate iv) still loaded - the viewer showed a blank empty_image.

- SecureIvRecovery.validatesImage: force a downsampled decode instead of a header-only inJustDecodeBounds check, so a wrong-iv candidate that only parses the header is rejected rather than persisted
- SecureRepository.load: re-check a 16-byte iv against the format magic bytes and re-recover if it's wrong, instead of trusting size==16
- HorizontalImageList: on secure full-image failure fall back to the working thumbnail rather than the blank empty_image drawable
- add SecureIvHeaderTest for the magic-byte logic

### CODE REVIEW

**Comment by tyrypyrking** [comment]
Tests are optional, if you see them as redundant we can scrap those - will lighten the PR by half.

**Comment by tyrypyrking** [comment]
My primary suspects for data corruption root causes are data races and thread safety. Things that are easy to run issues with when not using *the language that starts with R*.

File: app/src/main/java/com/kaii/photos/repositories/SecureRepository.kt [hunk 1]
```diff
@@ -286,7 +286,12 @@ class SecureRepository(
                 // pad corrupted/short ivs to 16 zero bytes so the [fileIv(16)][thumbnailIv(16)][path]
                 // layout holds and getThumbnailIv() doesn't read into the path bytes; recover first
                 val fileIv =
-                    if (iv != null && iv.size == 16) iv
+                    if (iv != null && iv.size == 16) {
+                        // a 16-byte iv can still be wrong (decrypts to garbage); re-check it against the
+                        // format magic and re-recover if bad rather than trusting size==16 forever
+                        if (SecureIvRecovery.ivProducesValidHeader(file, iv, mimeType)) iv
+                        else SecureIvRecovery.recoverAndPersist(context, file, mimeType, secureDao) ?: iv
+                    }
```
Comment (CompeyDev) [inline app/src/main/java/com/kaii/photos/repositories/SecureRepository.kt:294]: Previously, this used to be a zero-cost check, but now every image has 32 bytes read and a new `Cipher` is initialized. This should probably be memoized.

File: app/src/main/java/com/kaii/photos/helpers/SecureIvRecovery.kt [hunk 2]
```diff
@@ -141,24 +141,85 @@ object SecureIvRecovery {
...
+        return true
+    }
+
+    /** Invariant leading bytes per gated format, or null when we don't gate the format. */
+    fun magicFor(mimeType: String?): ByteArray? {
+        val m = mimeType?.lowercase() ?: return null
+        return when {
+            m.contains("png") -> PNG_MAGIC
+            m.contains("jpeg") || m.contains("jpg") -> JPEG_MAGIC
+            else -> null
```
Comment (CompeyDev) [inline app/src/main/java/com/kaii/photos/helpers/SecureIvRecovery.kt:199]: What about videos, webp, and gifs?

It returns `null` for these, and `ivProducesValidHeader` propagates this as `true` (meaning the IV doesn't need to be checked), although there is already code to recover IVs from videos.

File: app/src/main/java/com/kaii/photos/helpers/SecureIvRecovery.kt [hunk 3]
```diff
@@ -141,24 +141,85 @@ object SecureIvRecovery {
...
+
+    /** Pure: does [firstBlock] begin with every byte of [magic]? Exposed for unit testing. */
+    fun headerMatchesMagic(firstBlock: ByteArray, magic: ByteArray): Boolean {
+        if (firstBlock.size < magic.size) return false
+        for (i in magic.indices) if (firstBlock[i] != magic[i]) return false
+        return true
+    }
+
+    /** Invariant leading bytes per gated format, or null when we don't gate the format. */
+    fun magicFor(mimeType: String?): ByteArray? {
```
Comment (CompeyDev) [inline app/src/main/java/com/kaii/photos/helpers/SecureIvRecovery.kt:191]: Small nitpick, but since this was made public for the tests, we probably shouldn't be returning the direct instances to the magic signatures. This lets a caller mutate the globally shared instances. 

Consider using `.copyOf()` on them.

File: app/src/test/java/com/kaii/photos/SecureIvHeaderTest.kt
Comment (CompeyDev) [inline app/src/test/java/com/kaii/photos/SecureIvHeaderTest.kt:1]: It would be nice to have a test case for `computeInSampleSize`. It is pure arithmetic, so this should be possible.

I can think of a few edge cases (early short-circuiting, 2s growth, etc.) that we should test for.

**Review by CompeyDev — CHANGES_REQUESTED** [review]
The general logic looks good. There are some small edge-cases that need to be addressed, however.

**Comment by tyrypyrking** [comment]
Good suggestions. I'll be quite busy in coming weeks, so don't expect rapid responses from me.

File: app/src/main/java/com/kaii/photos/repositories/SecureRepository.kt (hunk 1 above)
Comment (tyrypyrking) [reply app/src/main/java/com/kaii/photos/repositories/SecureRepository.kt:294]: Memoized — verified paths are cached in a `verifiedIvs` set, so a passing iv isn't re-read or re-decrypted on reprocessing.

File: app/src/main/java/com/kaii/photos/helpers/SecureIvRecovery.kt (hunk 2 above)
Comment (tyrypyrking) [reply app/src/main/java/com/kaii/photos/helpers/SecureIvRecovery.kt:199]: `magicFor` now gates webp (RIFF), gif (GIF8) and webm (EBML), so a wrong iv for those routes into recovery; mp4/mov stay ungated since they have no byte-0 magic.

File: app/src/main/java/com/kaii/photos/helpers/SecureIvRecovery.kt (hunk 3 above)
Comment (tyrypyrking) [reply app/src/main/java/com/kaii/photos/helpers/SecureIvRecovery.kt:191]: Fixed — `magicFor` now returns `.copyOf()` so the shared constants can't be mutated through it.

File: app/src/test/java/com/kaii/photos/SecureIvHeaderTest.kt
Comment (tyrypyrking) [reply app/src/test/java/com/kaii/photos/SecureIvHeaderTest.kt:1]: Added — `computeInSampleSize` is now `internal` with tests for the at/below-target short-circuit, power-of-two growth, smaller-dimension gating, and zero/tiny dims.

File: app/src/main/java/com/kaii/photos/helpers/SecureIvRecovery.kt (hunk 2 above)
Comment (tyrypyrking) [reply app/src/main/java/com/kaii/photos/helpers/SecureIvRecovery.kt:199]: Other possible format are an open debate, but this is already stretching it. I'd rather have a different solution than an ever-growing list of magics to check against. Not a good maintenance strategy, imho.

Comment (CompeyDev) [reply app/src/main/java/com/kaii/photos/helpers/SecureIvRecovery.kt:199]: I'm not very familiar with Android dev / kotlin, but perhaps we could pull in a dependency like libmagic for this. I think you have a point. 

But that is a future concern not do with this PR.

**Review by CompeyDev — APPROVED** [review]
Looks good to me!
